### PR TITLE
Extend HFEM Kernels and BCs to support normal variables

### LINKED ISSUE #17747: Extend HFEM Kernels and BCs to support normal variables
## Reason

The existing HFEM (hybrid FEM) kernels and associated BCs (#17448) for were only works with array variables. To support normal variables, the existing ones were extended. 

## Design

Adopted the existing code design of HFEM for array variables was adopted and ported to normal variables.

## Impact

HFEM kernels and BCs can support usage of both normal and array variables.

### PR DESCRIPTION
**Pull request by yjung-anl**
closes  #17747

### CODE REVIEW

File: framework/doc/content/source/dgkernels/ArrayHFEMDiffusion.md
Comment (YaqiWang) [inline framework/doc/content/source/dgkernels/ArrayHFEMDiffusion.md:8]: Are we missing `.md` here? We will see when civet posts the link for the updated webpages of this PR.

File: test/tests/kernels/hfem/robin.i
Comment (YaqiWang) [inline test/tests/kernels/hfem/robin.i]: We do not need this aux variable and the `CoupledForce` kernel can be replaced with `BodyForce` kernel.

File: test/tests/kernels/hfem/neumann.i
Comment (YaqiWang) [inline test/tests/kernels/hfem/neumann.i]: We do not need this aux var if we replace `CoupledForce` with `BodyForce` kernel.

File: test/tests/kernels/hfem/dirichlet.i
Comment (YaqiWang) [inline test/tests/kernels/hfem/dirichlet.i]: We do not this aux var if we use `BodyForce` instead of `CoupledForce` kernel.

File: framework/src/dgkernels/DGLowerDKernel.C
Comment (YaqiWang) [inline framework/src/dgkernels/DGLowerDKernel.C:246]: This does not seem right. Possibly you want to check out `Kernel::computeOffDiagJacobian` for handling coupling array variables into a normal kernel.

File: framework/src/dgkernels/DGLowerDKernel.C
Comment (YaqiWang) [inline framework/src/dgkernels/DGLowerDKernel.C:251]: array DG kernel => DG kernel.

File: framework/src/dgkernels/DGLowerDKernel.C
Comment (YaqiWang) [reply framework/src/dgkernels/DGLowerDKernel.C:246]: But looks like `DGKernel::computeOffDiagElemNeighJacobian` is not doing things right either. Maybe we can just throw an error in this `if (jvar.fieldType() == Moose::VarFieldType::VAR_FIELD_ARRAY)` branch. @lindsayad do you have suggestions?

File: framework/src/bcs/LowerDIntegratedBC.C
Comment (YaqiWang) [inline framework/src/bcs/LowerDIntegratedBC.C:181]: This is similar as the DGKernel.

File: framework/src/bcs/HFEMDirichletBC.C
Comment (YaqiWang) [inline framework/src/bcs/HFEMDirichletBC.C:88]: One formatting suggestion: We can remove `{}` here in this if branch because there is only one statement in this branch. There are some other places in this PR where we can remove `{}` as well.

File: test/tests/kernels/hfem/dirichlet.i
Comment (YaqiWang) [inline test/tests/kernels/hfem/dirichlet.i:22]: Looks like we do not have any kernels for `lambdab` so we are not really testing the other way of imposing Dirichlet BCs. I suggest to either add the kernels for `lambdab` or remove this variable completely.

File: framework/include/bcs/HFEMDirichletBC.h
Comment (YaqiWang) [inline framework/include/bcs/HFEMDirichletBC.h]: Possibly we will get documentation errors because we are missing few markdown files for the newly added dg kernels and bcs.

**Review by YaqiWang — CHANGES_REQUESTED** [review]
Looks good overall. One big comment is the off-diagonal jacobian assembly with coupled array variables.

**Comment by YaqiWang** [comment]
Ah, I reviewed too fast. Hopefully my comments are still valid. We will need @lindsayad to have a second eye on this.

File: framework/doc/content/source/dgkernels/ArrayHFEMDiffusion.md
Comment (yjung-anl) [reply framework/doc/content/source/dgkernels/ArrayHFEMDiffusion.md:8]: I think it should be " Its standard variable version is [HFEMDiffusion.md]"

File: framework/src/dgkernels/DGLowerDKernel.C
Comment (yjung-anl) [reply framework/src/dgkernels/DGLowerDKernel.C:246]: Basically, I made this part based on `DGKernel::computeOffDiagElemNeighJacobian`. Can you give me more specific guide to correct it?

Comment (YaqiWang) [reply framework/src/dgkernels/DGLowerDKernel.C:246]: `Kernel::computeOffDiagJacobian` should be the place to check.

Comment (yjung-anl) [reply framework/src/dgkernels/DGLowerDKernel.C:246]: Know I see. You are expecting to have something like this:
`DGLowerDKernel::computeOffDiagLowerDJacobian`
```C
      for (_i = 0; _i < test_space.size(); _i++)
        for (_j = 0; _j < loc_phi.size(); _j++)
        {
         RealEigenVector v = _JxW[_qp] * _coord[_qp] * computeLowerDQpOffDiagJacobianArray(type, jvar);
         _assembly.saveFullLocalArrayJacobian(
              _local_ke, _i, test_space.size(), _j, loc_phi.size(), ivar, jvar.number(), v);
        }
```

`HFEMDirichletBC::computeLowerDQpOffDiagJacobianArray`
```C
RealEigenMatrix
HFEMDirichletBC::computeLowerDQpOffDiagJacobianArray(Moose::ConstraintJacobianType type,
                                                     const MooseVariableFEBase & jvar)
{
  if (_uhat_var && jvar.number() == _uhat_var->number() && type == Moose::LowerLower)
    return -_phi[_j][_qp] * _test_lambda[_i][_qp];
  else
    return RealEigenMatrix::Zero(_count, jvar.count());
}
```

File: test/tests/kernels/hfem/dirichlet.i
Comment (yjung-anl) [reply test/tests/kernels/hfem/dirichlet.i:22]: `lambdab` is used in here:
```java
[BCs]
  [all]
    type = HFEMDirichletBC
    boundary = 'left right top bottom'
    variable = u
    lowerd_variable = lambdab
  []
[]
```
Do you expect me to delete this one?

Comment (YaqiWang) [reply test/tests/kernels/hfem/dirichlet.i:22]: instead of a lowerd_variable here, we can directly use the given value. Essentially, lambdab is a primal variable, but no equation is for it currently.

File: framework/src/dgkernels/DGLowerDKernel.C
Comment (YaqiWang) [reply framework/src/dgkernels/DGLowerDKernel.C:246]: Yes. But we may not add this and just throw an error. We can add it later if needs come in the future.

Comment (yjung-anl) [reply framework/src/dgkernels/DGLowerDKernel.C:246]: I updated to just throw an error.

**Comment by yjung-anl** [comment]
Yaqi, do you have any idea how to activate civet testing? I have the following message: 

Test debug Pending — Developer needed to activate

**Comment by YaqiWang** [comment]
No, but I have asked moose guys. Hopefully they will answer you soon.

**Comment by roystgnr** [comment]
I just hit the "Activate all jobs" button; I'm not sure what the requirements are to get you added to the list of users whose PRs auto-activate so I'll leave that question to others.

File: test/tests/kernels/hfem/dirichlet.i
Comment (yjung-anl) [reply test/tests/kernels/hfem/dirichlet.i:22]: I added additional kernels to `lambdab`: 
```
  [reaction]
    type = Reaction
    variable = uhat
    rate = '1'
    block = BOUNDARY_SIDE_LOWERD_SUBDOMAIN
    reaction_coefficient = rc
  []
  [uhat_coupled]
    type = CoupledForce
    variable = uhat
    block = BOUNDARY_SIDE_LOWERD_SUBDOMAIN
    v = lambdab
    coef = '1'
  []
```

File: framework/src/bcs/LowerDIntegratedBC.C
Comment (YaqiWang) [inline framework/src/bcs/LowerDIntegratedBC.C:170]: DG kernel => integrated BC

File: framework/src/bcs/LowerDIntegratedBC.C
Comment (YaqiWang) [inline framework/src/bcs/LowerDIntegratedBC.C:172]: Same here.

File: test/tests/kernels/hfem/array_dirichlet.i
Comment (YaqiWang) [inline test/tests/kernels/hfem/array_dirichlet.i:70]: replace this with `value = '0 0'`.

File: test/tests/kernels/hfem/array_dirichlet.i
Comment (YaqiWang) [inline test/tests/kernels/hfem/array_dirichlet.i:24]: We can remove this variable. Possibly we will need a trivial regolding.

**Comment by YaqiWang** [comment]
Besides a few minor comments, this PR looks good to me.

File: test/tests/kernels/hfem/array_dirichlet.i
Comment (yjung-anl) [reply test/tests/kernels/hfem/array_dirichlet.i:70]:
```
[BCs]
  [all]
    type = ArrayHFEMDirichletBC
    boundary = 'left right top bottom'
    variable = u
    value = '0 0'
  []
[]
```
With this modification, the test failed with the following error:

tests/kernels/hfem/array_dirichlet.i:66: missing required parameter 'BCs/all/lowerd_variable'
        Doc String: "Lagrange multiplier"

**Comment by lindsayad** [comment]
Once Yaqi approves of what's been done, I will review (and probably
approve).

**Comment by yjung-anl** [comment]
> I just hit the "Activate all jobs" button; I'm not sure what the requirements are to get you added to the list of users whose PRs auto-activate so I'll leave that question to others.

**Comment by lindsayad** [comment]
Why was this closed?

**Comment by yjung-anl** [comment]
Oh It is my mistake.. accidentally hit the "close"

File: test/tests/kernels/hfem/array_dirichlet.i
Comment (YaqiWang) [reply test/tests/kernels/hfem/array_dirichlet.i:70]: oh, sorry. My bad. This is indeed needed and we need this variable as well.

File: test/tests/kernels/hfem/dirichlet.i
Comment (YaqiWang) [inline test/tests/kernels/hfem/dirichlet.i:62]: this parameter is not valid for Reaction?

**Comment by YaqiWang** [comment]
I only have one comment left now. Looks like that invalid parameter causing some test targets fail. The failure in Griffin can be patched later.

File: test/tests/kernels/hfem/dirichlet.i
Comment (yjung-anl) [reply test/tests/kernels/hfem/dirichlet.i:62]: Oh. I deleted

**Comment by YaqiWang** [comment]
@lindsayad I think you can start reviewing now.

File: test/tests/kernels/hfem/array_dirichlet.i
Comment (YaqiWang) [reply test/tests/kernels/hfem/array_dirichlet.i:24]: Discard this comment.

File: framework/include/dgkernels/DGLowerDKernel.h
Comment (lindsayad) [inline framework/include/dgkernels/DGLowerDKernel.h:14]: Do not need this

File: framework/src/bcs/LowerDIntegratedBC.C
Comment (lindsayad) [inline framework/src/bcs/LowerDIntegratedBC.C:55]:
```suggestion
      _local_re(_i) += _JxW[_qp] * _coord[_qp] * computeLowerDQpResidual();
```
For single instructions, let's remove surrounding curly braces

File: framework/src/bcs/LowerDIntegratedBC.C
Comment (lindsayad) [inline framework/src/bcs/LowerDIntegratedBC.C:86]:
```suggestion
  const auto jvar = (type == Moose::LowerLower || type == Moose::PrimaryLower)
```

Can make this `const`

File: framework/src/bcs/LowerDIntegratedBC.C
Comment (lindsayad) [inline framework/src/bcs/LowerDIntegratedBC.C:108]:
```suggestion
        _local_ke(_i, _j) += _JxW[_qp] * _coord[_qp] * computeLowerDQpJacobian(type);
```

File: framework/src/bcs/LowerDIntegratedBC.C
Comment (lindsayad) [inline framework/src/bcs/LowerDIntegratedBC.C:143]:
```suggestion
  const auto ivar = (type == Moose::LowerLower || type == Moose::LowerPrimary)
```

File: framework/src/bcs/LowerDIntegratedBC.C
Comment (lindsayad) [inline framework/src/bcs/LowerDIntegratedBC.C:166]:
```suggestion
          _local_ke(_i, _j) += _JxW[_qp] * _coord[_qp] * computeLowerDQpOffDiagJacobian(type, jvar);
```

File: framework/src/dgkernels/DGLowerDKernel.C
Comment (lindsayad) [inline framework/src/dgkernels/DGLowerDKernel.C:77]:
```suggestion
      _local_re(_i) += _JxW[_qp] * _coord[_qp] * computeLowerDQpResidual();
```

File: framework/src/dgkernels/DGLowerDKernel.C
Comment (lindsayad) [inline framework/src/dgkernels/DGLowerDKernel.C:118]:
```suggestion
  const auto & test_space =
```

File: framework/src/dgkernels/DGLowerDKernel.C
Comment (lindsayad) [inline framework/src/dgkernels/DGLowerDKernel.C:122]:
```suggestion
  const auto ivar =
```

File: framework/src/dgkernels/DGLowerDKernel.C
Comment (lindsayad) [inline framework/src/dgkernels/DGLowerDKernel.C:127]:
```suggestion
  const auto & loc_phi =
```

File: framework/src/dgkernels/DGLowerDKernel.C
Comment (lindsayad) [inline framework/src/dgkernels/DGLowerDKernel.C:131]:
```suggestion
  const auto jvar =
```

File: framework/src/dgkernels/DGLowerDKernel.C
Comment (lindsayad) [inline framework/src/dgkernels/DGLowerDKernel.C:148]:
```suggestion
        _local_ke(_i, _j) += _JxW[_qp] * _coord[_qp] * computeLowerDQpJacobian(type);
```

File: framework/src/dgkernels/DGLowerDKernel.C
Comment (lindsayad) [inline framework/src/dgkernels/DGLowerDKernel.C:201]:
```suggestion
  const auto & test_space =
```

File: framework/src/dgkernels/DGLowerDKernel.C
Comment (lindsayad) [inline framework/src/dgkernels/DGLowerDKernel.C:205]:
```suggestion
  const auto ivar =
```

File: framework/src/dgkernels/DGLowerDKernel.C
Comment (lindsayad) [inline framework/src/dgkernels/DGLowerDKernel.C:229]:
```suggestion
          _local_ke(_i, _j) += _JxW[_qp] * _coord[_qp] * computeLowerDQpOffDiagJacobian(type, jvar);
```

File: test/tests/kernels/hfem/tests
Comment (lindsayad) [inline test/tests/kernels/hfem/tests:48]:
```suggestion
    requirement = 'The system shall support hybrid finite element method (HFEM) calculations with Robin boundary conditions using array variables.'
```

File: test/tests/kernels/hfem/tests
Comment (lindsayad) [inline test/tests/kernels/hfem/tests:32]:
```suggestion
    requirement = 'The system shall support hybrid finite element method (HFEM) calculations with Neumann boundary conditions using array variables.'
```

**Review by lindsayad — CHANGES_REQUESTED** [review]
Overall LGTM

**Comment by yjung-anl** [comment]
Thank you for your review. I will updates my commits

**Comment by yjung-anl** [comment]
Updated commits were pushed

**Comment by yjung-anl** [comment]
Resolved the issue of activating the civet testing. After setting my two-factor auth on my github account, testings are up-and-running now.

**Comment by lindsayad** [comment]
I believe they only ran because I activated them. We have to add you to the moose collaborators GitHub group in order for your testing to be automatic. We will do that soon

**Comment by lindsayad** [comment]
@YaqiWang does someone have the griffin patch ready for this?

**Comment by loganharbour** [comment]
> We have to add you to the moose collaborators GitHub group in order for your testing to be automatic. We will do that soon

Added

**Comment by YaqiWang** [comment]
I suppose @yjung-anl will have one soon. The failure in one test is caused by renaming of one object and should be very quick to fix. I would not hold off merging this.

**Review by lindsayad — APPROVED** [review]

**Comment by lindsayad** [comment]
> I would not hold off merging this.

This will block any other content from getting from next to devel so I *am* going to hold off merging this until I know a fix is ready for Griffin. If it is an easy fix, that's good, then someone can ping me about a ready fix soon.

**Comment by yjung-anl** [comment]
@YaqiWang Perhaps can you share the error message of test failures? I can not see error message of these two tests:

`You are not allowed to see results. Please sign in to an appropiate user`

**Comment by YaqiWang** [comment]
@yjung-anl You can compile Griffin, run the tests and will see the error for now. I will ask MOOSE guys to set up this properly for you. It might take some time because you may not have access to some other controlled apps.

**Comment by YaqiWang** [comment]
@yjung-anl You can patch Griffin now. In that patch, you will need to do a manual MOOSE submodule update with its hash to the current next branch. Once the MR is in, I will merge it asap.

**Pull request merged**
